### PR TITLE
Add pdo/cards scripts, script infrastructure changes

### PR DESCRIPTION
**Pull request by ThirteenFish**
This adds
 - The `pdo` script, which can list or listen for TPDOs from specified cards
 - The `cards` script, which lists available cards for a given Oresat mission
 - Improvements to the top level `oresat_configs` script dispatcher and some options to make discovery easier

### CODE REVIEW

File: oresat_configs/__init__.py
Comment (dmitri-mcguckin) [inline oresat_configs/__init__.py:45]: Please add either safeguards in code or corresponding docs as to what the structure of the CSV is expected to be.

File: oresat_configs/__init__.py
Comment (dmitri-mcguckin) [inline oresat_configs/__init__.py:63]: I would prefer this stay as `oresat_id` so we have the added context as to what this is.

That or add a pydoc string describing it. Either or.

File: oresat_configs/__main__.py
Comment (dmitri-mcguckin) [inline oresat_configs/__main__.py:49]: LGTM

We've been under-utilizing argparses abilities up until now.

File: oresat_configs/_yaml_to_od.py
Comment (dmitri-mcguckin) [inline oresat_configs/_yaml_to_od.py:675]: Same deal as earlier, please either add context to the var name or add a docstring describing it.

File: oresat_configs/scripts/gen_dcf.py
Comment (dmitri-mcguckin) [inline oresat_configs/scripts/gen_dcf.py:14]: Is there a reason this exists as duplicate code across all of the OD sub-parsers?

Could this be moved to the root level of the cmd line tool?

I don't forsee a use case where we built an xtce generator but not the DCF generator.

File: oresat_configs/scripts/gen_fw_files.py
Comment (dmitri-mcguckin) [inline oresat_configs/scripts/gen_fw_files.py:16]: Same deal as `gen_dcf.build_parser` file. See that comment for details.

File: oresat_configs/scripts/gen_xtce.py
Comment (dmitri-mcguckin) [inline oresat_configs/scripts/gen_xtce.py:15]: Same deal as `gen_dcf.build_parser` file. See that comment for details.

File: oresat_configs/scripts/pdo.py
Comment (dmitri-mcguckin) [inline oresat_configs/scripts/pdo.py:105]: Maybe I'm not familiar with this enough, but is there something this is doing that we're not already achieving with SavvyCAN?

I was hoping to get away from any freewheeled CAN parsers across OreSat projects, since keeping a parser up-to-date is a lot of work.

This also seems out of scope from what configs was intended to be: a simple static config generator/manager.

File: pyproject.toml
Comment (dmitri-mcguckin) [inline pyproject.toml:69]: Do we have documentation/justification for adding `W0511`?

Adding this to your local project is fine but I think it would be best-practice to make sure we're not pushing/keeping `TODO`'s in mainline.

**Review by dmitri-mcguckin — CHANGES_REQUESTED** [review]
A couple of notes:

* The parser building in each generator I'm not quite understanding, could we maybe have one parser for the tool and have each generator add a subparser type instead?
* I'd like to make the `oresat` object in the Configs have a bit more context to it than just `oresat`, it was previously `oresat_id` is it no longer representing an ID?

Other than that looks good. It's been in need of some argparse spruce up for a while now.

File: oresat_configs/__init__.py
Comment (ThirteenFish) [reply oresat_configs/__init__.py:45]: It'll `KeyError` when it doesn't find the necessary columns. Are you asking for a more descriptive error? I'm unsure what you're intending by safeguard here.

The structure of the CSV is expected to be whatever is in `cards.csv`. As far as I can tell there should only ever be one of those. If a column is removed it'll `KeyError`, if a column is added, you won't find it in the related `Card` until you explicitly add it to `Card`, and `Card` won't let you construct itself until it gets that new column.

Comment (dmitri-mcguckin) [reply oresat_configs/__init__.py:45]: Right, that's what worries me- some time down the line we update `cards.csv` and it'll start throwing key errors thus crashing any generators.

Simply allowing KeyErrors to pass is not graceful IMO and likely to be a source of frustration down the road. We should either throw a useful error like *"Cards.csv is missing \_ expected columns"** or *"Has \_ unexpected column"* or something.

File: oresat_configs/scripts/gen_dcf.py
Comment (ThirteenFish) [reply oresat_configs/scripts/gen_dcf.py:14]: `pyproject.toml` installs each subscript as both a standalone script and as a subscript to the root level `oresat-configs` (the `[project.scripts]` section). Because they need to be able to run without the root level tool, they need to construct their own arguments on their own. We could have another file `script-args.py` that held the shared code and have both the root level and subscripts import from that, but a bit of copy/pasting seemed simple enough.

For example, I very much wanted to put the `--oresat` argument in the root level because every script takes it. If I were to then run `oresat-gen-dcf` though, the standalone version, it wouldn't have the `--oresat` option because it bypassed the root level.

If we're OK with not having the standalone versions of the scripts, that would allow the implementation to de-duplicate a bunch of the arguments up into the root level. I didn't get rid of them here to keep backwards compatibility.

File: pyproject.toml
Comment (ThirteenFish) [reply pyproject.toml:69]: I'm going to go with it's perverse incentive. I personally use TODOs and FIXMEs for things things that are worth noting, but not so much that they're worth writing a ticket over, and also not specific to the current commit/pull request I'm working on, or are enough effort that they'd be good in their own commit.

I agree that having TODOs and FIXMEs accumulate is not the best practice, they should be attended to eventually, but I think failing CI on them is worse. I was hoping that I could convince `pylama` to emit a warning and not have it count as a fail, but I didn't see how to convince it to do so.

The perverse incentive is that if this fails CI, I will continue to write these notes but now not label them with TODO or FIXME, making them harder to find. Or just not return to the feature. Both those options seem worse.

File: oresat_configs/scripts/pdo.py
Comment (ThirteenFish) [reply oresat_configs/scripts/pdo.py:105]: It's not a CAN parser, at least not directly. It goes through CANopen, specifically avoiding raw CAN messages. It assembles itself directly from the Object Dictionary generically, so there's not much to keep up-to-date. When the yaml configs get updated it gets updated automatically. As far as I can tell SavvyCAN isn't CANopen aware, the user needs to add their own definitions on it. 

My plan is to enhance this to be able to send as well, to a given card's RPDOs, and eventually add in RTR PDOs as well. Unless I missed something, SavvyCAN is capture only? 

As far as scope, this is the dual to `transfer-sdo.py`, which is in this project. Both the PDO and SDO scripts could be broken out into a separate repo, but that would be a small repo and the Oresat project already suffers from too many of them.

File: oresat_configs/__init__.py
Comment (ThirteenFish) [reply oresat_configs/__init__.py:63]: A docsting is probably the better choice here. It can't stay as an `oresat_id` because it's generally no longer just an ID, it's a `Consts`. As described in the messages for [63f8766](https://github.com/oresat/oresat-configs/pull/16/commits/63f8766959d5665f5c01e731b8f7b1bb373443a6) and [777c060](https://github.com/oresat/oresat-configs/pull/16/commits/777c0602a7f53b146c0e30a1554229c0743d0628), `Consts` has the ID, but also has other information related to a specific mission, currently long and short names, and paths to config files as well.

I kept the variable name `oresat` because then all the uses can be changed from `oresat_id` -> `oresat.id` as the `id` member is where the information now resides. If you've got ideas for the name (for both the variable and the type) I'd love to hear them though and am open to changing them. As the saying goes, the two hardest things in programming are cache invalidation, naming things, and off by one errors.

File: pyproject.toml
Comment (dmitri-mcguckin) [reply pyproject.toml:69]: I'd argue that you can and should be able to continue to use `TODO`'s and all the other useful labels- so long as they are in a stub branch off-mainline and that anything with `TODO`-attached code should be bisected out of the to-merge branch before merging the to-merge branch into main.

I get where you're coming from, I just think there's a way to keep those labels for dev and not prod- is all.

Happy to hear other thoughts and opinions too.

@ryanpdx  what do you think of a *"no todo's in main"*-esqe policy for prod repos?

Comment (ryanpdx) [reply pyproject.toml:69]: Please do not use `TODO` or `FIXME`, they lead to technical debt

File: oresat_configs/scripts/pdo.py
Comment (ryanpdx) [inline oresat_configs/scripts/pdo.py:67]: We do not use unicode str, int24, or uin24, remove them.

File: oresat_configs/__main__.py
Comment (ryanpdx) [inline oresat_configs/__main__.py:22]: Remove TODO comment

File: oresat_configs/__main__.py
Comment (ryanpdx) [inline oresat_configs/__main__.py:9]: Could be apart of the file docstring

**Review by ryanpdx — COMMENTED** [review]
Just minor things, otherwise looks good.

File: pyproject.toml
Comment (dmitri-mcguckin) [reply pyproject.toml:69]: I think it's fine to use them- so long as they're all resolved before merge.

If they aren't resolved- then why? Is it part of a feature that should be its own PR? Is it a task item that should be cleaned up first? etc.

File: oresat_configs/scripts/pdo.py
Comment (ThirteenFish) [reply oresat_configs/scripts/pdo.py:67]: This dict is for display names of standard CANopen types (at least as implemented through the `canopen` python package). This does not use them in the configs, and the configs can continue not using those types. However, not including those types here would increase the possible valid PDO mappings that this script could not not display correctly. It would reduce compliance with the CANopen standard.

File: pyproject.toml
Comment (ThirteenFish) [reply pyproject.toml:69]: > Please do not use `TODO` or `FIXME`, they lead to technical debt

They are a symptom not a cause, right? Cart-before-the-horse and all that. Technical debt is caused by lack of $resource during development. A lack of time, of knowledge, of design and planning, of stable requirements, and so on. Some debt is known at the time of writing but can't or won't be addressed due to the aforementioned $resource shortage. Why would we not want that explicitly called out in the code? The author is best positioned to identify these known instances, why would we want to discourage them from telling others about that? The tech debt will be there whether they comment on it or not.

Good comments answer the question why. Why is the code written the way that it is? What was going through the author's mind, what was their intent when writing this piece of code? `TODO`s and `FIXME`s can fit in that category because they answer the question "why not?". The author didn't check this invariant, or implement this in the way I would have expected, or left out this seemingly obvious part of a feature because they lacked $resource, not because it was an intentional design choice. I as someone visiting the codebase later can have more confidence in the changes I make knowing I'm helping the author instead of accidentally stepping on their toes. 

Code comments explicitly are a good way of marking this because of locality. As the codebase evolves, since the comments are right there, they have the best chance of getting updated and staying relevant. If the note lives far away from the code it will drift more. It's not obvious from the code itself that there might be a ticket attached to that specific line, or that there's a branch off from 245 commits back that references it.

Adding the annotation makes the reviewer's job easier too by letting the areas that specifically need more review stand out. Reviewers can help resolve them if they find the issue easier (that they have access to $resource that the submitter did not), or just generally be kept aware of what to look out for in the future.

> I think it's fine to use them- so long as they're all resolved before merge.
> 
> If they aren't resolved- then why? Is it part of a feature that should be its own PR? Is it a task item that should be cleaned up first? etc.

This is effectively the current policy, CI is only run on pushes or merge requests to master, not on branches. The challenge is that code with `TODO`s attached to them is still useful, it accomplishes the task set out for the PR sufficiently, but it hasn't quite lived up to the author's vision. I agree that asking why they weren't resolved is worthwhile, and reviewers can help solve them, but sometimes the answer genuinely is some issues can't or shouldn't be resolved now. For example: waiting on an upstream library to release a new version to use a feature from it. There's nothing the author can do about that but wait. (if you're meticulous you can spot at least one instance of this in this very PR, but mea culpa I didn't mark it).

Code goes to die on stale feature branches. It's hard enough convincing people to read code on master, not a single person will read the stub branch besides the author.  Stale branches can drift quickly from master without active maintenance, merges can get hairy quickly. When revisiting the issue, now not only are there the concerns of the issue itself, there's the extra barrier of fighting through a merge on top. On master the comments get to transform with the code, it's always merged and up to date.

Another benefit of `TODO`s is they're good first project candidates. A new codebase is intimidating, and these provide clear points of study for people looking to contribute. New people are typically only going to look through master.

Comment (ryanpdx) [reply pyproject.toml:69]: Well, I disagree with everything you just said. I have dealt with too many codebases full of `FIXME`s, `TODO`s, and commented out code where everyone is unsure if they are still relevant. 

But we have no coding standards and this PR is not the place to discuss/decide on them, so I will approve it as everything looks good.

**Review by ryanpdx — APPROVED** [review]

**Pull request merged**
